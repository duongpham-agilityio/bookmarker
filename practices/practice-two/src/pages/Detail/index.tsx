import { useContext, useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import { useBook } from 'hooks';

// HOCs
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

// Contexts
import { FormContext } from 'contexts/Form/context';

// Components
import { Button, Heading } from 'components/commons';
import { Error } from 'components';
import DetailSkeleton from 'pages/Detail/Skeleton';

// Styles
import containerStyles from 'styles/commons/index.module.css';
import styles from 'pages/Detail/index.module.css';

// Assets
import TrashIcon from 'assets/icons/trash.svg';
import BackIcon from 'assets/icons/back.svg';
import PenCilIcon from 'assets/icons/pencil.svg';
import { convertDateTimeToTimeString, convertTimeToDate } from 'helpers';

const Detail = () => {
  const { data, error, isLoading, deleteBook } = useBook();
  const redirect = useNavigate();

  if (error) {
    return <Error />;
  }

  const { dispatch } = useContext(FormContext);
  const book = useMemo(() => {
    const { publishDate, createdAt, updatedAt, ...rest } = data;
    const publishDateConvert = `${convertDateTimeToTimeString(
      publishDate
    )}, ${convertTimeToDate(publishDate, '/', true)}`;
    const createdAtConvert = `${convertDateTimeToTimeString(
      createdAt
    )}, ${convertTimeToDate(createdAt, '/', true)}`;
    const updatedAtConvert = `${convertDateTimeToTimeString(
      updatedAt
    )}, ${convertTimeToDate(updatedAt, '/', true)}`;

    return {
      ...rest,
      publishDate: publishDateConvert,
      createdAt: createdAtConvert,
      updatedAt: updatedAtConvert,
    };
  }, [data]);

  return (
    <main className={containerStyles.container}>
      {isLoading ? (
        <DetailSkeleton />
      ) : (
        <section className={styles.detail}>
          <div className={styles.detailItem}>
            <Heading label={data.name} size="xl" className={styles.heading} />
            <div>
              <div className={styles.action}>
                <Button
                  label=""
                  leftIcon={TrashIcon}
                  variant="danger"
                  onClick={deleteBook}
                />
                <Button
                  label=""
                  leftIcon={BackIcon}
                  variant="primary"
                  onClick={() => redirect('/books')}
                />
                <Button
                  label="Edit"
                  leftIcon={PenCilIcon}
                  variant="primary"
                  width="w-lg"
                  onClick={() =>
                    dispatch({
                      formData: data,
                      title: 'Edit book',
                      type: 'update',
                    })
                  }
                />
              </div>
            </div>
          </div>
          <div className={styles.detailItem}>
            <div className={styles.info}>
              <p className={styles.description}>
                <span>{data.description.substring(0, 150)}</span>
                <span className={styles.actionText}>show more</span>
              </p>
              <ul className={styles.listInfo}>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Author: </span>
                  <span className={styles.text}>{book.author}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Published date: </span>
                  <span className={styles.text}>{book.publishDate}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Created at: </span>
                  <span className={styles.text}>{book.createdAt}</span>
                </li>
                <li className={styles.infoItem}>
                  <span className={styles.label}>Updated at: </span>
                  <span className={styles.text}>{book.updatedAt}</span>
                </li>
              </ul>
            </div>
            <div className={styles.imageWrap}>
              <img
                src={data.imageURL}
                alt={data.name}
                className={styles.image}
              />
            </div>
          </div>
        </section>
      )}
    </main>
  );
};

export default withErrorBoundaries(Detail);
