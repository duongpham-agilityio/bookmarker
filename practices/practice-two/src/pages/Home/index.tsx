import { ChangeEvent, MouseEvent, useContext, useState } from 'react';
import { Link } from 'react-router-dom';
import { useBooks, useDebounce } from 'hooks';

// HOCS
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

// Contexts
import { FormContext } from 'contexts/Form/context';

//Components
import { Button, Heading, Input } from 'components/commons';
import { Card, CardSkeleton, Error } from 'components';

// Constants
import { ENDPOINT, MESSAGES, SORT, TITLE_FORM } from '@constants';

// Styles
import homeStyles from 'pages/Home/index.module.css';
import commonStyles from 'styles/commons/index.module.css';

// Assets
import SearchIcon from 'assets/icons/search.svg';
import AddIcon from 'assets/icons/add.svg';

// Helpers
import { convertDateTimeToTimeString } from 'helpers';

const Home = () => {
  const { dispatch } = useContext(FormContext);
  const {
    param: { page: currentPage, sort },
    data: dataShow,
    isLoading,
    error,
    pagination,
    deleteBook,
    setSearchParam,
    convertSearchParamsToString,
  } = useBooks();
  const [search, setSearch] = useState('');
  const debounce = useDebounce((value) => setSearchParam('name', value));

  if (error) {
    return <Error />;
  }

  return (
    <main className={commonStyles.container}>
      <section className={homeStyles.home}>
        <div className={homeStyles.navbar}>
          <Input
            className={homeStyles.search}
            value={search}
            leftIcon={SearchIcon}
            placeholder="Search something..."
            onChange={(event: ChangeEvent<HTMLInputElement>) => {
              setSearch(event.target.value);
              debounce(event.target.value);
            }}
          />
          <nav className={homeStyles.navigation}>
            <ul className={homeStyles.navList}>
              <li>
                <Link
                  to={`/${ENDPOINT.BOOKS}${convertSearchParamsToString(
                    'sort',
                    SORT.ASCENDING
                  )}`}
                  className={`${homeStyles.navLink} ${
                    sort === SORT.ASCENDING ? homeStyles.active : ''
                  }`}
                >
                  Ascending
                </Link>
              </li>
              <li>
                <Link
                  to={`/${ENDPOINT.BOOKS}${convertSearchParamsToString(
                    'sort',
                    SORT.DESCENDING
                  )}`}
                  className={`${homeStyles.navLink} ${
                    sort === SORT.DESCENDING ? homeStyles.active : ''
                  }`}
                >
                  Descending
                </Link>
              </li>
            </ul>
          </nav>
          <Button
            label="Create"
            variant="primary"
            leftIcon={AddIcon}
            width="w-lg"
            border="b-lg"
            onClick={() => {
              dispatch({
                formData: {
                  author: '',
                  description: '',
                  imageURL: '',
                  name: '',
                  createdAt: new Date().getTime(),
                  deletedAt: null,
                  updatedAt: new Date().getTime(),
                },
                title: TITLE_FORM.CREATE,
                type: 'create',
              });
            }}
          />
        </div>

        {isLoading ? (
          <div className={homeStyles.grid}>
            {[1, 2, 3, 4, 5, 6].map((value) => (
              <CardSkeleton key={value} />
            ))}
          </div>
        ) : (
          <>
            {dataShow?.length ? (
              <>
                <div className={homeStyles.content}>
                  <div className={homeStyles.grid}>
                    {dataShow.map((book) => (
                      <Card
                        href={`/${ENDPOINT.BOOKS}/${book.id}`}
                        title={book.name}
                        description={book.description}
                        publishedDate={convertDateTimeToTimeString(
                          book.createdAt
                        )}
                        imageUrl={book.imageURL}
                        key={book.id}
                        onDelete={(event: MouseEvent) => {
                          event.preventDefault();

                          deleteBook(book.id || 0);
                        }}
                      />
                    ))}
                  </div>
                </div>
                <div className={homeStyles.pagination}>
                  {pagination.length > 1 &&
                    pagination.map((__, index) => {
                      const page = index + 1;

                      return (
                        <Button
                          label={`${page}`}
                          variant={
                            currentPage === page ? 'primary' : 'secondary'
                          }
                          size="small"
                          key={index}
                          onClick={() => {
                            setSearchParam('page', `${page}`);
                          }}
                        />
                      );
                    })}
                </div>
              </>
            ) : (
              <div className={homeStyles.empty}>
                <Heading
                  label={MESSAGES.EMPTY_TITLE}
                  className={homeStyles.emptyTitle}
                  size="xl"
                />
                <p className={homeStyles.emptyDescription}>
                  {MESSAGES.EMPTY_DESCRIPTION}
                </p>
              </div>
            )}
          </>
        )}
      </section>
    </main>
  );
};

export default withErrorBoundaries(Home);
