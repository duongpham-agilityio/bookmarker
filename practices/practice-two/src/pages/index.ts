import { lazy } from 'react';

const Home = lazy(() => import('pages/Home'));
const Detail = lazy(() => import('pages/Detail'));

export { Home, Detail };
