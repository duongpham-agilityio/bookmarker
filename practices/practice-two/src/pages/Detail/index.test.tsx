import { act, fireEvent, render } from '@testing-library/react';
import { Suspense } from 'react';

// Pages
import { Detail } from 'pages';
import { BrowserRouter } from 'react-router-dom';

const setup = async () => {
  return await act(() =>
    render(
      <Suspense>
        <BrowserRouter>
          <Detail />
        </BrowserRouter>
      </Suspense>
    )
  );
};

jest.mock('assets/icons/x-letter.svg', () => ({
  default: 'assets/icons/x-letter.svg',
}));

jest.mock('assets/icons/upload.svg', () => ({
  default: 'assets/icons/upload.svg',
}));

jest.mock('assets/icons/next.svg', () => ({
  default: 'assets/icons/next.svg',
}));

jest.mock('assets/icons/trash.svg', () => ({
  default: 'assets/icons/trash.svg',
}));

jest.mock('assets/icons/back.svg', () => ({
  default: 'assets/icons/back.svg',
}));

jest.mock('assets/icons/pencil.svg', () => ({
  default: 'assets/icons/pencil.svg',
}));

jest.mock('hooks', () => ({
  useBook: jest.fn().mockReturnValue({
    data: {
      name: 'HTML/CSS Ebook 1',
      description: 'Description of some book will displayed here',
      author: 'Duong.Pham',
      imageURL:
        'http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api',
      createdAt: 1680158351376,
      deletedAt: null,
      updatedAt: 1680158351376,
      publishDate: 1680158351376,
      id: 1,
    },
  }),
}));

describe('Details', () => {
  it('match snapshot', async () => {
    const { asFragment } = await setup();

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', async () => {
    const { findByText } = await setup();

    expect(await findByText(/HTML\/CSS Ebook 1/i)).toBeInTheDocument();
  });

  it('action', async () => {
    const mockDispatcher = jest.fn();
    const { findByText } = await setup();
    const button = await findByText(/edit/i);

    button.addEventListener('click', mockDispatcher);

    act(() => {
      fireEvent.click(button);
    });

    expect(mockDispatcher).toHaveBeenCalled();
  });
});
