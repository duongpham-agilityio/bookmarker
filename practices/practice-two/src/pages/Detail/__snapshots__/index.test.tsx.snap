// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details match snapshot 1`] = `
<DocumentFragment>
  <main>
    <section>
      <div>
        <h2
          class="undefined undefined "
        >
          HTML/CSS Ebook 1
        </h2>
        <div>
          <div>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="icon action"
                src="assets/icons/trash.svg"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="icon action"
                src="assets/icons/back.svg"
              />
            </button>
            <button
              class="undefined undefined
  undefined undefined
  undefined "
              type="button"
            >
              <img
                alt="icon action"
                src="assets/icons/pencil.svg"
              />
              Edit
            </button>
          </div>
        </div>
      </div>
      <div>
        <div>
          <p>
            <span>
              Description of some book will displayed here
            </span>
          </p>
          <ul>
            <li>
              <span>
                Author: 
              </span>
              <span>
                Duong.Pham
              </span>
            </li>
            <li>
              <span>
                Published date: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
            <li>
              <span>
                Created at: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
            <li>
              <span>
                Updated at: 
              </span>
              <span>
                1:39 PM, 30/03/2023
              </span>
            </li>
          </ul>
        </div>
        <div>
          <img
            alt="HTML/CSS Ebook 1"
            src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
